### PR TITLE
Feature/s3 c 3265 warp10 failover

### PR DESCRIPTION
**Pull request by tmacro**

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -9,8 +9,25 @@
         "dumpLevel": "error"
     },
     "redis": {
-        "host": "127.0.0.1",
-        "port": 6379
+        "name": "scality-s3",
+        "password": "",        "sentinels": [
```
Comment (miniscruff) [inline config.json]: weird formatting, maybe hit backspace one too many times?

File: libV2/config/index.js [hunk 2]
```diff
@@ -1,14 +1,25 @@
 const fs = require('fs');
 const path = require('path');
 const Joi = require('@hapi/joi');
+const asert = require('assert');
 
 const { truthy, envNamespace } = require('../constants');
 const configSchema = require('./schema');
 
+function _splitServer(text) {
+    assert(text.indexOf(':') !== -1);
```
Comment (miniscruff) [inline libV2/config/index.js]: I feel like assert strict not equal would be more appropriate

File: config.json (hunk 1 above)
Comment (tmacro) [reply config.json]: whoops that wasn't supposed to be staged

**Review by miniscruff — APPROVED** [review]
Did not find anything that looked wrong minus the formatting stuff, overall looks good.

File: libV2/config/schema.js [hunk 3]
```diff
@@ -16,6 +16,24 @@ const redisSentinelSchema = Joi.object({
     sentinelPassword: Joi.string().default('').allow(''),
 });
 
+const warp10SingleHost = Joi.object({
+    host: Joi.alternatives(Joi.string().hostname(), Joi.string().ip()),
```
Comment (jonathan-gramain) [inline libV2/config/schema.js:20]: Maybe you could add `.required()` to `host` and `port` if that is the case

File: libV2/config/schema.js [hunk 4]
```diff
@@ -16,6 +16,24 @@ const redisSentinelSchema = Joi.object({
...
+    port: Joi.number().port(),
+    readToken: Joi.string(),
+    writeToken: Joi.string(),
+});
+
+const warp10MultiHost = Joi.object({
+    hosts: Joi.array().items(Joi.object({
```
Comment (jonathan-gramain) [inline libV2/config/schema.js:27]: same for `.required()` to `hosts`

File: libV2/warp10.js [hunk 5]
```diff
@@ -15,11 +21,36 @@ class Warp10Client {
         this._readToken = (config && config.readToken) || 'readTokenCI';
         this._nodeId = (config && config.nodeId) || _config.nodeId;
         const proto = (config && config.tls) ? 'https' : 'http';
-        const host = (config && config.host) || 'localhost';
-        const port = (config && config.port) || 4802;
         const requestTimeout = (config && config.requestTimeout) || 10000;
         const connectTimeout = (config && config.connectTimeout) || 10000;
-        this._warp10 = new Warp10(`${proto}://${host}:${port}`, requestTimeout, connectTimeout);
+        if (config && config.hosts) {
+            this._clients = config.hosts.map(h => {
```
Comment (jonathan-gramain) [inline libV2/warp10.js]: I think this arrow function could be written as `({ host, port }) => new Warp10(...)`

File: libV2/warp10.js [hunk 6]
```diff
@@ -15,11 +21,36 @@ class Warp10Client {
...
+    }
+
+    async _wrapCall(func, params) {
+        // eslint-disable-next-line no-restricted-syntax
+        for (const client of this._clients) {
+            try {
+                // eslint-disable-next-line no-await-in-loop
+                return await func(client, ...params);
+            } catch (error) {
+                moduleLogger.debug('error during warp10 operation, failing over to next host', { error });
```
Comment (jonathan-gramain) [inline libV2/warp10.js]: could be a `warn` log instead?

**Review by vrancurel — APPROVED** [review]

**Pull request merged**